### PR TITLE
Change initial status of sofort donations

### PR DESCRIPTION
**Pull request by gbirke**
Sofort donations now have an initial status of "X" (unconfirmed external
donation) and get a status of "Z" (payment promise) after they are
confirmed.

This avoids unconfirmed & bogus donations from being exported.

Previously, the status of sofort donations was always "Z" (payment
promise) and the moderation status was never set because
notifyOfPolicyValidationFailure only sets the status for non-external
payments.

This is for fixing https://phabricator.wikimedia.org/T179468

### CODE REVIEW

File: src/Domain/Model/Donation.php [hunk 1]
```diff
@@ -199,6 +199,11 @@ public function confirmBooked(): void {
 			$this->makeCommentPrivate();
 		}
 
+		if ( $this->getPaymentType() === PaymentType::SOFORT ) {
+			$this->status = self::STATUS_PROMISE;
+			return;
+		}
+
```
Comment (JeroenDeDauw) [inline src/Domain/Model/Donation.php]: My first thought here is that this is misplaced because it makes Donation bind to a specific payment method. Will have a more detailed look at the code later and hopefully find an alternative

Comment (JeroenDeDauw) [reply src/Domain/Model/Donation.php]: I'm ok with merging this as-is to get the issue resolved and then following up.

After looking at the code a bit, it seems like `confirmBooked` might belong in the DonationPayment rather than in the whole Donation.

Except for the `makeCommentPrivate` part, which has me slightly confused. It looks to be doing "if a donation had a non published status before, and it was in moderation or canceled, make the comment private before we put it into a published state". If that is the case, why can't the comment be marked private as soon as the thing is put into moderation or canceled? Are there other cases in which the comment should end up being public?

When moving this to the DonationPayment class, the `hasExternalPayment` bit becomes more cohesive. The opposite happens for `statusAllowsForBooking`, which makes me wonder about the status field. It seems like most values are about payment status, with moderation being an exception and perhaps cancellation as well. Could this be pulled apart? (You can look at membership Application... I'm not sure the same approach would work here due to lack of domain knowledge.)

Having the sofort specific stuff in DonationPayment would still be weird. Could move it further into PaymentMethod by adding a method like `canConfirmAsBooked` to the interface that does the same check as the current `hasExternalPayment` bit. Though if the status that results from confirming something as booked is not always the same, you'd need to have a method that takes the current status and returns the new one instead.

Is it really needed to have the same action result in different status based on the payment type? Somehow this makes me think something went wrong somewhere.

Comment (gbirke) [reply src/Domain/Model/Donation.php]: IMHO what went wrong is that status field has evolved/mutated over the history of the application to indicate three different things:

* **Payment status:** This is only needed for payments that have status changes, i.e. external payments where the donation is written and later confirmed by a notification from the payment provider. Before sofort, we had two states for external payments, "new" (`X`) and "booked" (`B`), and two superfluous additional states that mirror the payment type (`N` for direct debit and `Z` for bank transfer. Confirmed Sofort donations now also have `Z` because Sofort can't guarantee that we will get the money, so Grün needs an indication that this is a "promised payment" like a bank tranfer. When we redo this, we could live with two states and handle special case of Sofort in the export, because it's really for Grün.
* **General status**: Donations can be "new", "exported" and "deleted". Currently,  "deleted" is a status `X`. Deletion and the other states are represented by the respective timestamp field being `NULL` or having a timestamp. 
* **Flagging for Moderation**: "Bad Words" in address fields or an amount being too big trigger a flag where the fundraising team has to look at the donation in the Fundraising Operation Center to either clear the flag or delete the donation. When a donation gets a flag, its comment is automatically set to private.

If we don't do radical changes to the underlying data model ( *\*cough\** Eventsourcing \*cough\**), we could at least use different database fields for the three different concerns.

Comment (JeroenDeDauw) [reply src/Domain/Model/Donation.php]: > When we redo this, we could live with two states and handle special case of Sofort in the export, because it's really for Grün.

Why can't we put this in the export stuff now?

> If we don't do radical changes to the underlying data model ( *cough* Eventsourcing *cough**), we could at least use different database fields for the three different concerns.

The domain model can be cleaned up without touching the db (and is more important than the db... esp if the db is private to the data access layer which it is not yet due to FOC app). This is what we did for membership applications, see DoctrineApplicationRepository::getDoctrineStatus

Comment (gbirke) [reply src/Domain/Model/Donation.php]: > Why can't we put this in the export stuff now?

Now that I've written about the states, it seems like a good idea to do this. The only argument against it is that it'll take more time and simultaneous deployments of frontend and backend. All the while bogus donations are coming in ...

> The domain model can be cleaned up without touching the db

That would result in two methods in the data access layer, with lots of `if` statements to convert the 3 status from the domain into 1 db status and back. And lots of coordination with Till for the `convertDomainStatesToDBStatus` method on what status (payment, general, moderation) trumps what status and in what case. I'd rather change the data model.

**Comment by JeroenDeDauw** [comment]
> Do we migrate existing SUB donations so we have a consistent DB?

We agreed to move the status change stuff to the export, so stuff in the db will not change and remain consistent

File: src/Domain/Model/Donation.php (hunk 1 above)
Comment (JeroenDeDauw) [reply src/Domain/Model/Donation.php]: Gabriel and I agreed to do the if check in the export as this turns out to be easy (1h kinda task), which avoids the payment type problem I commented on here. I will do that this weekend. The PR can already be merged as is (and I prefer if this is done). Only reason I'm not merging already is the first of Pablo's questions which I can't answer myself.

**Comment by gbirke** [comment]
> Do we need to sync this with gruen?

TL;DR: No
The premise of these changes is that nothing will change in the export for Gruen. But when this change is merged, we won't export bogus donations to Gruen in the future.

> Do we migrate existing SUB donations so we have a consistent DB?

A tricky question: Do we want to have a consistent DB or do we want the to be DB consistent with what we exported to Gruen? I'd say the first, but would also ask @KaiNissen 

A migration script would do the following for each donation with type "SUB":
* Set status to "X" if Status is "Z" and confirmation date (in the data blob) is missing.
* Set status to "B" if Status is "Z" and confirmation date exists.

**Comment by JeroenDeDauw** [comment]
+2 to everything except my little commit at the end

**Comment by gbirke** [comment]
+2 for Jeroens commit.
